File: wlr/OGUR/OGUR/Items/GenericItem.cs
using System;
using System.Collections.Generic;
using OGUR.Collision;
using OGUR.Creatures;
using OGUR.GameObjects;
using OGUR.Management;
using OGUR.Sprites;

namespace OGUR.Items
{
    public class GenericItem : GameplayObject
    {
        public Stats Modifers;
        public String Name;
        private string m_suffix;
        private string m_prefix;
        private int m_type;
        private Slots m_targetSlots;

        private const string s_spacingCharacter = " ";

        public int GetItemClass()
        {
            switch (m_type)
            {
                case ItemName.Sword:
                    return ItemClass.Melee_Weapon;
                case ItemName.Arrow:
                    return ItemClass.Ranged_Ammo;
                case ItemName.Bow:
                    return ItemClass.Ranged_Weapon;
                case ItemName.Shield:
                    return ItemClass.Shield;
                case ItemName.Pants:
                    return ItemClass.Leggings;
                case ItemName.Dagger:
                    return ItemClass.Melee_Weapon;
                case ItemName.Staff:
                    return ItemClass.Melee_Weapon;
                case ItemName.Hood:
                    return ItemClass.Head_Gear;
                case ItemName.Shirt:
                    return ItemClass.Torso_Garb;
                default:
                    return ItemClass.NULL;
            }
        }

        private void Initialize(string suffix, string prefix,int type, Slots targetSlots,Stats modifiers,Point2 location)
        {
            Setup(location, type);
            m_suffix = suffix;
            m_prefix = prefix;
            m_type = type;
            m_targetSlots = GetSlotFromType(type);
            Name = (m_prefix == ItemPrefix.NULL ? String.Empty :  m_prefix + s_spacingCharacter) +
                    ItemName.Names[m_type] +
                   (m_suffix == ItemSuffix.NULL ? String.Empty : s_spacingCharacter + m_suffix);
            Modifers = new Stats(modifiers);
        }

        public GenericItem(GenericItem item,Point2 location)
        {
            Initialize(item.m_suffix, item.m_prefix, item.m_type,item.m_targetSlots, item.Modifers,location);
        }

        public GenericItem(Stats modifiers, string suffix, string prefix, int type, Point2 location,bool onGround = true)
        {
            if (type == ItemName.NULL)
            {
                throw new Exception("Invalid type NULL passed into the GenericItem factory!");
            }
            Initialize(m_suffix,m_prefix,type,GetSlotFromType(type),modifiers,location);
        }

        protected void Setup(Point2 location, int type)
        {
            Initialize(location, SpriteFromItem(type), GameObjectType.ITEM,Depth.Item);
        }

        private Player m_currentTarget;
        public override void Update()
        {
            base.Update();
            if (m_isOnBoard)
            {
                m_currentTarget = GameplayObjectManager.GetTouchingPlayer(this);
                if (m_currentTarget != null)
                {
                    if (m_currentTarget.IsInteracting())
                    {
                        m_currentTarget.PickupItem(this);
                    }
                }
            }
        }

        private SpriteType SpriteFromItem(int item)
        {
            return SpriteType.ITEM;
        }

        private Slots GetSlotFromType(int type)
        {
            switch (type)
            {
                case ItemName.Sword:
                    return new Slots(new List<ItemSlot>() {ItemSlot.RIGHT_HAND, ItemSlot.LEFT_HAND});
                case ItemName.Pants:
                    return new Slots(new List<ItemSlot>() {ItemSlot.LEGS});
                default:
                    return new Slots(new List<ItemSlot>());
            }
        }

        public override bool Equals(object obj)
        {
            if(obj.GetType()==typeof(GenericItem))
            {
                var gI = (GenericItem)obj;
                if (Name != gI.Name)
                {
                    return false;
                }
                foreach (string stat in StatType.Values)
                {
                    if (Modifers.Get(stat) != gI.Modifers.Get(stat))
                    {
                        return false;
                    }
                }    
                return true;
            }
            return false;
        }

        public override int GetHashCode()
        {
            return Name.GetHashCode() + Modifers.GetHashCode();
        }

        public static bool operator ==(GenericItem a, GenericItem b)
        {
            if (System.Object.ReferenceEquals(a, b))
            {
                return true;
            }

            if (((object) a == null) || ((object) b == null))
            {
                return false;
            }

            return a.Equals(b);
        }

        public static bool operator !=(GenericItem a, GenericItem b)
        {
            return !(a == b);
        }

        public float GetStatBonus(string stat)
        {
            return Modifers.Get(stat);
        }
    }
}

File: proj/src/aigilas/items/GenericItem.java
package aigilas.items;

import aigilas.creatures.Player;
import aigilas.creatures.StatType;
import aigilas.creatures.Stats;
import spx.bridge.ActorType;
import spx.bridge.DrawDepth;
import spx.bridge.EntityType;
import spx.core.Point2;
import spx.entities.Entity;
import spx.entities.EntityManager;
import spx.entities.IActor;

public class GenericItem extends Entity {
    public Stats Modifers;
    public String Name;
    private String _suffix;
    private String _prefix;
    private ItemName _type;
    private Slots _targetSlots;

    private static final String __spacingCharacter = " ";

    public ItemClass getItemClass() {
        return _type.Category;
    }

    private void initialize(String suffix, String prefix, ItemName type, Slots targetSlots, Stats modifiers, Point2 location) {
        setup(location, type);
        _suffix = suffix;
        _prefix = prefix;
        _type = type;
        _targetSlots = type.Slots;
        Name = (_prefix == null ? "" : _prefix + __spacingCharacter) + _type.toString() + (_suffix == null ? "" : __spacingCharacter + _suffix);
        Modifers = new Stats(modifiers);
    }

    public GenericItem(GenericItem item, Point2 location) {
        initialize(item._suffix, item._prefix, item._type, item._targetSlots, item.Modifers, location);
    }

    public GenericItem(Stats modifiers, String suffix, String prefix, ItemName type, Point2 location, boolean onGround) {
        if (type == null) {
            try {
                throw new Exception("Invalid type NULL passed into the GenericItem factory!");
            } catch (Exception e) {

                e.printStackTrace();
            }
        }
        initialize(_suffix, _prefix, type, type.Slots, modifiers, location);
    }

    public GenericItem(Stats modifiers, String suffix, String prefix, ItemName type, Point2 location) {
        this(modifiers, suffix, prefix, type, location, true);
    }

    protected void setup(Point2 location, ItemName type) {
        initialize(location, type.Sprite, EntityType.ITEM, DrawDepth.Item);
    }

    private Player _currentTarget;

    @Override
    public void update() {
        super.update();
        if (_isOnBoard) {
            IActor collider = EntityManager.getTouchingCreature(this);
            if (collider != null && collider.getActorType() == ActorType.PLAYER) {
                _currentTarget = (Player) collider;
                if (_currentTarget != null) {
                    if (_currentTarget.isInteracting()) {
                        _currentTarget.pickupItem(this);
                    }
                }
            }
        }
    }

    @Override
    public boolean equals(Object obj) {
        if (obj.getClass() == GenericItem.class) {
            GenericItem gI = (GenericItem) obj;
            if (Name != gI.Name) {
                return false;
            }
            for (StatType stat : StatType.values()) {
                if (Modifers.get(stat) != gI.Modifers.get(stat)) {
                    return false;
                }
            }
            return true;
        }
        return false;
    }

    @Override
    public int hashCode() {
        return Name.hashCode() + Modifers.hashCode();
    }

    public float getStatBonus(StatType stat) {
        return Modifers.get(stat);
    }
}
